PHPCSUtils/Utils/TypeString.php

```diff
@@ -0,0 +1,403 @@
+            && \preg_match('`^[^|&()?\s]+(\s*&\s*[^|&()?\s]+)+$`', $typeString) === 1;
+    }
+
+    /**
+     * Check if a type string represents a disjunctive normal form (DNF) type.
+     *
+     * This check for a strict
+     *
+     * @since 1.1.0
+     *
+     * @param string $typeString Type string.
+     *
+     * @return bool
+     */
+    public static function isDNF($typeString)
+    {
+        return \is_string($typeString)
+            && \strpos($typeString, '?') === false
+            && \strpos($typeString, '|') !== false
+            && \strpos($typeString, '&') !== false
+            && \strpos($typeString, '(') !== false
+            && \strpos($typeString, ')') !== false
+            // Now make sure that it is not a definitely invalid format.
+            && \preg_match(self::INVALID_DNF_REGEX, $typeString) !== 1;
+    }
+
+    /**
+     * Split a type string to its individual types and optionally normalize the case of the types.
+     *
+     * @since 1.1.0
+     *
+     * @param string $typeString Type to split.
+     * @param bool   $normalize  Whether or not to normalize the case of types.
+     *                           Defaults to true.
+     *
+     * @return array<string> List containing all seen types in the order they were encountered.
+     *
+     * @throws \PHP_CodeSniffer\Exceptions\RuntimeException If passed $typeString is not a string.
+     */
+    public static function toArray($typeString, $normalize = true)
+    {
+        if (\is_string($typeString) === false) {
+            throw new RuntimeException('The $typeString parameter must be passed a string');
+        }
+
+        if (\trim($typeString) === '') {
+            return [];
+        }
+
+        $addNull = false;
+        if ($typeString[0] === '?') {
+            $addNull    = true;
+            $typeString = \substr($typeString, 1);
+        }
+
+        $typeString = \preg_replace('`\s+`', '', $typeString);
+        $types      = \preg_split('`[|&()]+`', $typeString, -1, \PREG_SPLIT_NO_EMPTY);
+
+        // Normalize the types.
+        if ($normalize === true) {
+            $types = \array_map([__CLASS__, 'normalizeCase'], $types);
+        }
+
+        if ($addNull === true) {
+            \array_unshift($types, 'null');
+        }
+
+        return $types;
+    }
+
+    /**
+     * Split a type string to the unique types included and optionally normalize the case of the types.
+     *
+     * @since 1.1.0
+     *
+     * @param string $typeString Type to split.
+     * @param bool   $normalize  Whether or not to normalize the case of types.
+     *                           Defaults to true.
+     *
+     * @return array<string, string> Associative array with the unique types as both the key as well as the value.
+     *
+     * @throws \PHP_CodeSniffer\Exceptions\RuntimeException If passed $typeString is not a string.
+     */
+    public static function toArrayUnique($typeString, $normalize = true)
```

Comment:
@jrfnl please can you help me understand why this should be a separate method, rather than built-in functionality of `toArray()`? When would a consumer of this library want to have a list of types which *are not* unique?

Reply:
> When would a consumer of this library want to have a list of types which _are not_ unique?

The most typical case is when a sniff wants to flag all uses of a certain class/interface name. If the name is used twice in the type, like `(A&B)|(A&D)`, it should also be flagged twice.

More of an edge case would be an error prevention sniff flagging non-OO duplicates in a type, like https://3v4l.org/FFjpT, or OO type redundancy, like https://3v4l.org/1heQE.